### ISSUE TITLE
Rack-aware load balancing

### ISSUE THREAD

**Issue by piodul** [scylladb/scylla-rust-driver#604]
Current cloud providers tend to organize their resources in a two-tier hierarchy: physical datacenters are called _availability zones_ which are grouped into _regions_. Scylla/Cassandra snitches for AWS and GCE translate those terms as follows:

- _availability zone_ -> __rack__
- _region_ -> __DC__

We already have a DC aware policy which routes requests to the local DC first. However, in this case a __DC__ is not actually a physical datacenter and may consist of multiple _availability zones_. Traffic between _availability zones_ is expensive, both in terms of latency and money.

We should let users keep their traffic within the local _availability zone_, i.e. __rack__. There are two aspects of this:

- Enable our load balancing to prefer the local rack. In our current composable policy design we could introduce a "rack aware policy", similar to our current DC aware policy.
- Allow to prevent the driver from connecting to nodes from other racks. We already have the HostFilter mechanism and it should just be a matter of providing an implementation which filters by DC and rack.

References:
- https://github.com/gocql/gocql/pull/1565
- https://github.com/gocql/gocql/pull/1659

**Comment by avelanarius** [comment]
Yes, I’ll assign somebody, but the customer said that gocql is their priority (this part assigned to @Lorak-mmk to merge the upstream to our fork). The customer also mentioned the C++ driver, where we have two options:
- C++ bindings which also require https://github.com/scylladb/scylla-rust-driver/issues/462 for the bindings to be able to get “raw bytes” unparsed response from driver (required by the customer)
- C++ driver, requiring some help from @jul-stas for the release process

**Comment by mykaul** [comment]
@avelanarius - is this complete for 0.8? If not, please re-target to 0.8.1 or 0.9.
